src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +16,144 @@
-    SPELL_SILENCE           = 18327
-    //SPELL_POSSESS           = 17244
+    SPELL_BANSHEEWAIL           = 16565,
+    SPELL_BANSHEECURSE          = 16867,
+    SPELL_SILENCE               = 18327,
+    SPELL_POSSESS               = 17244,    // The charm
+    SPELL_POSSESSED             = 17246     // The damage buff
 };
 
-class boss_baroness_anastari : public CreatureScript
+enum BaronessAnastariEvents
 {
-public:
-    boss_baroness_anastari() : CreatureScript("boss_baroness_anastari") { }
-
-    CreatureAI* GetAI(Creature* creature) const override
-    {
-        return GetStratholmeAI<boss_baroness_anastariAI>(creature);
-    }
-
-    struct boss_baroness_anastariAI : public ScriptedAI
-    {
-        boss_baroness_anastariAI(Creature* creature) : ScriptedAI(creature)
-        {
-            Initialize();
-            instance = me->GetInstanceScript();
-        }
+    EVENT_SPELL_BANSHEEWAIL     = 1,
+    EVENT_SPELL_BANSHEECURSE    = 2,
+    EVENT_SPELL_SILENCE         = 3,
+    EVENT_SPELL_POSSESS         = 4,
+    EVENT_INVISIBLE             = 5,
+    EVENT_CHECK_POSSESSED       = 6
+};
 
-        void Initialize()
-        {
-            BansheeWail_Timer = 1000;
-            BansheeCurse_Timer = 11000;
-            Silence_Timer = 13000;
-            //Possess_Timer = 35000;
-        }
 
-        InstanceScript* instance;
+struct boss_baroness_anastari : public BossAI
+{
+    boss_baroness_anastari(Creature* creature) : BossAI(creature, TYPE_BARONESS) { }
 
-        uint32 BansheeWail_Timer;
-        uint32 BansheeCurse_Timer;
-        uint32 Silence_Timer;
-        //uint32 Possess_Timer;
+    EventMap _events;
+    ObjectGuid _possessedTargetGuid;
 
-        void Reset() override
+    void Reset() override
+    {
+        if (!me->IsVisible())
         {
-            Initialize();
+            if (Player* possessedTarget = ObjectAccessor::FindConnectedPlayer(_possessedTargetGuid))
+            {
+                if (possessedTarget->IsAlive())
+                {
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
+                    me->SetReactState(REACT_AGGRESSIVE);
+                    me->SetVisible(true);
+                }
+            }
         }
+        _events.Reset();
+    }
 
-        void JustEngagedWith(Unit* /*who*/) override
-        {
-        }
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.ScheduleEvent(EVENT_SPELL_BANSHEEWAIL, 1s);
+        _events.ScheduleEvent(EVENT_SPELL_BANSHEECURSE, 11s);
+        _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+        _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+    }
 
-        void JustDied(Unit* /*killer*/) override
-        {
-            instance->SetData(TYPE_BARONESS, IN_PROGRESS);
-        }
+    void JustDied(Unit* /*killer*/) override
+    {
+        me->GetInstanceScript()->SetData(TYPE_BARONESS, IN_PROGRESS);
```

Comment:
```
instance->SetData(TYPE_BARONESS, IN_PROGRESS);
```
was fine

Reply:
Oh right. Fair point. I didn't think about that earlier.